File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = [-1] # dummy index, to prevent stack from empty error
        heights.append(0) # dummy node in the tail
        max_area = 0
        for i, cur_height in enumerate(heights):
            while stack and heights[stack[-1]] > cur_height:
                height = heights[stack.pop()]
                width = i - stack[-1] - 1
                max_area = max(max_area, width * height)
            stack.append(i)
        return max_area

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& arr) {
        int n = arr.size();
        int maxArea = 0;
        stack<int>st;
        for (int i = 0; i < n; i++) {
            while (!st.empty() && arr[st.top()] > arr[i]) {
                int el = st.top();
                st.pop();
                int pse = st.empty() ? -1 : st.top();
                maxArea = max(maxArea, (i - pse - 1) * arr[el]);
            }
            st.push(i);
        }
        while(!st.empty()){
            int el = st.top();
            st.pop();
            int pse = st.empty() ? -1 : st.top();
            maxArea = max(maxArea,(n-pse-1)*arr[el]);
        }
        return maxArea;
    }
};